### DISCUSSION TITLE
[Feature request] Layer Control

### DISCUSSION THREAD

**Discussion by santoshgistto** [hm21/pro_image_editor#136] [General]
### Platforms

Android, iOS, Web, Windows, macOS, Linux

### Description

Would be nice if we have support something like below for layer control for Pain, Text and Image

https://github.com/hm21/pro_image_editor/assets/12577046/f5bc6237-9cf1-43bb-be67-ac95f577a39e

### Why

we can style and resize better way specially for Text and Paint

**Comment by hm21** [comment]
Okay, I will add it to my [roadmap](https://github.com/users/hm21/projects/7). 
I had anyway the plan to update the layer interaction more so that we can also interact the same way we can do it in the crop-rotate-editor. However, even though this feature looks simple, it requires many changes, so I don't know when I will have time to do it. That means maybe in a few days or weeks I will release a version with this feature. 
For the case you need it quickly, I recommend you to create a PR. In case you want to create a PR, please let me know before because then I will also post here a list with the files you need to edit, so you don't have to search long for the files you need to edit.

**Comment by santoshgistto** [comment]
Yes please post the list of files, I will give a try

**Comment by hm21** [comment]
Sure below the list.

- [Layer](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/layer/layer.dart): Update layer so that it has `scaleX` and `scaleY` and not just `scale`.
- [LayerInteractionHelperWidget](https://github.com/hm21/pro_image_editor/blob/stable/lib/widgets/layer_interaction_helper/layer_interaction_helper_widget.dart): Add your widgets here, as in your video, to show the user that they can drag to resize.
- [LayerInteraction](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/editor_configs/layer_interaction_configs.dart): Extend the configs where the user can enable/disable this feature.
- [ImageEditorCustomWidgets](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/custom_widgets/custom_widgets.dart) Add in the customWidgets a new class `CustomWidgetsLayerInteraction` which allows the user to add their own widgets.
- [MainEditorCallbacks](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/editor_callbacks/main_editor_callbacks.dart): Add new callbacks that are triggered when the user interacts.
- [Layer](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/layer/layer.dart): Update layer so that it has `scaleX` and `scaleY` and not just `scale`.
- [ImportStateHistory](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/import_export/import_state_history.dart): Update import state history that it's backward compatible with old `scale` parameter, which did not have `scaleX` and `scaleY`.  The new version you can add [here](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/import_export/utils/export_import_version.dart). 
- [ThemeLayerInteraction](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/theme/theme_layer_interaction.dart): Add configs that allow the user to apply a unique design.
- [I18nLayerInteraction](https://github.com/hm21/pro_image_editor/blob/stable/lib/models/i18n/i18n_layer_interaction.dart): If you have new tooltips, update the i18n file.
- [LayerWidget](https://github.com/hm21/pro_image_editor/blob/stable/lib/widgets/layer_widget.dart) This file contains the generated layer. Maybe you don't need to change anything.
- [LayerInteractionManager](https://github.com/hm21/pro_image_editor/blob/721f072730ca59e703aba3e9a8a688e471336d52/lib/modules/main_editor/utils/layer_interaction_manager.dart#L136-L189) Not necessary, but it shows you how to calculate the distance that will be required when the layer is rotated.

Please let me know if you have any questions :)

**Reply by santoshgistto** [reply]
Thanks @hm21  
Couldn't get chance to look into it again sorry for that 
Pushed my initial changes here, still pending 
https://github.com/santoshgistto/pro_image_editor/commit/b53be0f542cd08ec05a213666de5b645dd2b0bf1

![Screenshot 2024-06-28 at 10 53 16 AM](https://github.com/hm21/pro_image_editor/assets/12577046/d57cf24c-ef5b-4622-a5ad-9e7ab2ced4fb)

**Reply by hm21** [reply]
@santoshgistto  I just check your good. Looks good and I'm excited to see the final solution :)

Just as a small note, I know it's not done, but keep in mind file names in flutter should always be in snake case, so the file `PillsWidget.dart` must have the name `pills_widget.dart`. The problem is, if we do not change it, the package will get less pub points which you can see [here](https://pub.dev/packages/pro_image_editor/score). You can also check the workflow [here](https://github.com/santoshgistto/pro_image_editor/actions/runs/9714743104/job/26814522994) which runs every time you commit. It will fail and show you the issues that will downgrade the pub points. Alternatively, you can also just run `flutter analyze .` in your terminal.

**Comment by hm21** [comment]
Btw, I convert it to discussion that we can talk here where also other users can see is open and in progress by you.
